### PR TITLE
[WIP] add example to reproduce benchmarks from paper

### PR DESCRIPTION
**Pull request by aaronkl**
add example scripts that show how to reproduce the HNAS experiments from the original paper

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-797-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/3/index.html

File: examples/hnas/hpolib2_example.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+def make_benchmark(dataset_name='cifar10-valid'):
+    b = NasBench201BaseBenchmark(dataset_name)
+    cs = b.get_configuration_space()
+
+    d = dict()
+    for h in cs.get_hyperparameters():
+        d[h.name] = ag.space.Categorical(*h.choices)
+
+    @ag.args(**d, epochs=199)
+    def objective_function(args, reporter, **kwargs):
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:28]: You can drop this **kwargs

File: examples/hnas/hpolib2_example.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        ts_start = time.time()
+
+        config = dict()
+        for h in cs.get_hyperparameters():
+            config[h.name] = args[h.name]
+
+        for epoch in range(args.epochs):
+
+            res = b.objective_function(config, fidelity={'epoch': epoch})
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:38]: Is this going to sleep for the time training would have taken? Otherwise, eval_time will be wrong

File: examples/hnas/hpolib2_example.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                epoch=epoch + 1,
+                performance=acc,
+                eval_time=eval_time,
+                time_step=ts_now, **config)
+
+    return objective_function
+
+
+# CLI
+def parse_args():
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:54]: At some point, we should centralize this stuff somewhere

File: examples/hnas/hpolib2_example.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                        help='Asynchronous scheduler type. In case of doubt leave it to the default')
+    args = parser.parse_args()
+    return args
+
+
+if __name__ == "__main__":
+
+    args = parse_args()
+
+    logging.root.setLevel(logging.INFO)
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:81]: This did not work for me.

File: examples/hnas/resnet_example.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+                        help='number of GPUs available to a given trial.')
+    parser.add_argument('--num_cpus', type=int, default=4,
+                        help='number of CPUs available to a given trial.')
+    parser.add_argument('--hostfile', type=argparse.FileType('r'))
+    parser.add_argument('--store_results_period', type=int, default=100,
+                        help='If specified, results are stored in intervals of '
+                             'this many seconds (they are always stored at '
+                             'the end)')
+    parser.add_argument('--scheduler', type=str, default='hyperband_promotion',
+                        choices=['hyperband_stopping', 'hyperband_promotion'],
```
Comment (mseeger) [inline examples/hnas/resnet_example.py:257]: Why not 'fifo' as well?

File: examples/hnas/hpolib2_example.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                        help='number of GPUs available to a given trial.')
+    parser.add_argument('--num_cpus', type=int, default=2,
+                        help='number of CPUs available to a given trial.')
+
+    parser.add_argument('--store_results_period', type=int, default=100,
+                        help='If specified, results are stored in intervals of '
+                             'this many seconds (they are always stored at '
+                             'the end)')
+    parser.add_argument('--scheduler', type=str, default='hyperband_promotion',
+                        choices=['hyperband_stopping', 'hyperband_promotion'],
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:71]: Why not 'fifo' as well?

File: examples/hnas/hpolib2_example.py (hunk 3 above)
Comment (mseeger) [reply examples/hnas/hpolib2_example.py:54]: In fact, we might want to go back to the previous setup, where one could select the train_fn function.

File: examples/hnas/resnet_example.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+                        help='Asynchronous scheduler type. In case of doubt leave it to the default')
+    args = parser.parse_args()
+    return args
+
+
+if __name__ == "__main__":
+
+    args = parse_args()
+
+    logging.root.setLevel(logging.INFO)
```
Comment (mseeger) [inline examples/hnas/resnet_example.py:267]: Did not work for me

File: examples/hnas/resnet_example.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+                                                time_out=args.timeout,
+                                                # This argument defines the metric that will be maximized.
+                                                # Make sure that you report this back in the objective function.
+                                                reward_attr='performance',
+                                                # The metric along we make scheduling decision. Needs to be also
+                                                # reported back to AutoGluon in the objective function.
+                                                time_attr='epoch',
+                                                brackets=brackets,
+                                                checkpoint=None,
+                                                searcher="bayesopt",  # Defines searcher for new configurations
```
Comment (mseeger) [inline examples/hnas/resnet_example.py:301]: Why not also allow 'random' here?

File: examples/hnas/hpolib2_example.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                                                time_out=args.timeout,
+                                                # This argument defines the metric that will be maximized.
+                                                # Make sure that you report this back in the objective function.
+                                                reward_attr='performance',
+                                                # The metric along we make scheduling decision. Needs to be also
+                                                # reported back to AutoGluon in the objective function.
+                                                time_attr='epoch',
+                                                brackets=brackets,
+                                                checkpoint=None,
+                                                searcher="bayesopt",  # Defines searcher for new configurations
```
Comment (mseeger) [inline examples/hnas/hpolib2_example.py:105]: Why not also allow 'random' here?

File: examples/hnas/hpolib2_example.py (hunk 1 above)
Comment (aaronkl) [reply examples/hnas/hpolib2_example.py:28]: done

File: examples/hnas/hpolib2_example.py (hunk 2 above)
Comment (aaronkl) [reply examples/hnas/hpolib2_example.py:38]: you're right, I have added an option to put the worker asleep

File: examples/hnas/hpolib2_example.py (hunk 4 above)
Comment (aaronkl) [reply examples/hnas/hpolib2_example.py:81]: I've put it on the top of the file

File: examples/hnas/resnet_example.py (hunk 5 above)
Comment (aaronkl) [reply examples/hnas/resnet_example.py:257]: ok I added it

File: examples/hnas/hpolib2_example.py (hunk 6 above)
Comment (aaronkl) [reply examples/hnas/hpolib2_example.py:71]: ok I added it

File: examples/hnas/resnet_example.py (hunk 7 above)
Comment (aaronkl) [reply examples/hnas/resnet_example.py:267]: I've put it on the top of the file

File: examples/hnas/resnet_example.py (hunk 8 above)
Comment (aaronkl) [reply examples/hnas/resnet_example.py:301]: done

File: examples/hnas/hpolib2_example.py (hunk 9 above)
Comment (aaronkl) [reply examples/hnas/hpolib2_example.py:105]: done

**Review by mseeger — APPROVED** [review]

**Comment by szha** [comment]
Job PR-797-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/6/index.html

File: examples/hnas/lstm_example.py [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+        data = 'wikitext-2/'
+
+        emsize = 1500
+        nhid = 1500
+        nlayers = 2
+        eval_batch_size = 10
+        bptt = 35
+
+        tied = True
+        seed = np.random.randint(10000)
```
Comment (gradientsky) [inline examples/hnas/lstm_example.py:270]: Do we want to fix seed if it is provided in args? Otherwise this one will harm reproducibility.

File: examples/hnas/lstm_example.py [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+        clip = args.clip
+        lr_factor = args.lr_factor
+
+        ts_start = time.time()  # Time stamp for elapsed_time
+
+        torch.manual_seed(seed)
+        if torch.cuda.is_available():
+            device = torch.device("cuda")
+        else:
+            device = torch.device("cpu")
```
Comment (gradientsky) [inline examples/hnas/lstm_example.py:287]: Consider adding other 'flavors' of fixed seed:
```python
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)  # if you are using multi-GPU.
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True
```

File: examples/hnas/resnet_example.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+
+    def forward(self, x):
+        x = self.resnet(x)
+        x = F.log_softmax(x, dim=1)
+
+        return x
+
+
+def train(model, train_loader, optimizer, epoch):
+    model.train()
```
Comment (gradientsky) [inline examples/hnas/resnet_example.py:74]: Add `model.cuda()`

File: examples/hnas/resnet_example.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+
+    total_loss = []
+
+    for data, target in train_loader:
+        if torch.cuda.is_available():
```
Comment (gradientsky) [inline examples/hnas/resnet_example.py:79]: Cache value so it's not queried on every batch?

File: examples/hnas/resnet_example.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+        weight_decay=ag.space.Real(lower=1e-5, upper=1e-3, log=True),
+        lr=ag.space.Real(lower=1e-3, upper=1e-1, log=True),
+        dataset_path=dataset_path,
+        num_gpus=num_gpus,
+        epochs=27)
+    def objective_function(args, reporter):
+
+        ts_start = time.time()
+
+        torch.manual_seed(np.random.randint(10000))
```
Comment (gradientsky) [inline examples/hnas/resnet_example.py:142]: ditto - seed comment

**Comment by szha** [comment]
Job PR-797-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/8/index.html

**Comment by szha** [comment]
Job PR-797-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/9/index.html

**Comment by szha** [comment]
Job PR-797-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/10/index.html

**Comment by szha** [comment]
Job PR-797-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/11/index.html

**Comment by szha** [comment]
Job PR-797-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/12/index.html

**Comment by szha** [comment]
Job PR-797-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/13/index.html

**Comment by szha** [comment]
Job PR-797-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/14/index.html

**Comment by szha** [comment]
Job PR-797-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-797/15/index.html

**Comment by Innixma** [comment]
Please re-open if this is still relevant, closing due to inactivity.
